gk/fib.c

```diff
@@ -63,40 +64,83 @@ static inline int
 initialize_ether_cache(struct ether_cache *eth_cache)
 {
 	memset(eth_cache, 0, sizeof(*eth_cache));
+	seqlock_init(&eth_cache->lock);
 
 	return 0;
 }
 
-/*
- * Fill up the Ethernet cached header.
- * Note that the destination MAC address should be filled up by LLS.
- */
-static inline void
+static void
+gk_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_cache *eth_cache = arg;
+
+	/*
+	 * Deal with concurrency control by sequential lock
+	 * on the nexthop entry.
+	 */
+	write_seqlock(&eth_cache->lock);
+
+	if (map->stale) {
+		ether_addr_copy(&map->ha, &eth_cache->eth_hdr.d_addr);
+		eth_cache->stale = false;
+	}
```

Comment:
Shouldn't we perform an update even if the map is not stale? For example, if the update is to make the entry go from non-stale to stale, or to change the L2 address of a non-stale entry.